atom/browser/ui/win/thumbar_host.cc

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2015 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "atom/browser/ui/win/thumbar_host.h"
+
+#include <shobjidl.h>
+
+#include "base/win/scoped_comptr.h"
+#include "base/win/win_util.h"
+#include "base/win/wrapped_window_proc.h"
+#include "base/strings/utf_string_conversions.h"
+#include "third_party/skia/include/core/SkBitmap.h"
+#include "ui/gfx/icon_util.h"
+#include "ui/gfx/win/hwnd_util.h"
+
+namespace atom {
+
+namespace {
+
+// The thumbnail toolbar has a maximum of seven buttons due to the limited room.
+const int kMaxButtonsCount = 7;
+
+bool GetThumbarButtonFlags(const std::vector<std::string>& flags,
+                           THUMBBUTTONFLAGS* out) {
+  if (flags.empty()) {
+    *out = THBF_ENABLED;
+    return true;
+  }
+  THUMBBUTTONFLAGS result = static_cast<THUMBBUTTONFLAGS>(0);
+  for (const auto& flag : flags) {
+    if (flag == "enabled") {
+      result |= THBF_ENABLED;
+    } else if (flag == "disabled") {
+      result |= THBF_DISABLED;
+    } else if (flag == "dismissonclick") {
+      result |= THBF_DISMISSONCLICK;
+    } else if (flag == "nobackground") {
+      result |= THBF_NOBACKGROUND;
+    } else if (flag == "hidden") {
+      result |= THBF_HIDDEN;
+    } else if (flag == "noninteractive") {
+      result |= THBF_NONINTERACTIVE;
+    } else {
+      return false;
+    }
+  }
+  *out = result;
+  return true;
+}
+
+}  // namespace
+
+ThumbarHost::ThumbarHost(HWND window) : is_initialized_(false),
+                                        window_(window) {
+}
+
+ThumbarHost::~ThumbarHost() {
+}
+
+bool ThumbarHost::SetThumbarButtons(
+    const std::vector<ThumbarHost::ThumbarButton>& buttons) {
+  THUMBBUTTON thumb_buttons[kMaxButtonsCount];
+  thumbar_button_clicked_callback_map_.clear();
+
+  // Once a toolbar with a set of buttons is added to thumbnail, there is no way
+  // to remove it without re-creating the window.
+  // To achieve to re-set thumbar buttons,  we initialize the buttons with
+  // HIDDEN state and only updated the caller's specified buttons.
+  //
+  // Initialize all thumb buttons with HIDDEN state.
+  for (int i = 0; i < kMaxButtonsCount; ++i) {
+    thumb_buttons[i].iId = i;
```

Comment:
The buttons ID should start with a relatively large number like 100, otherwise it is going to have same ID with built-in command IDs.